pkg/util/util.go

```diff
@@ -0,0 +1,159 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"strings"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/kylelemons/godebug/diff"
+)
+
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
+var (
+	// debugPackage controls verbose debugging in this package. Used for offline debugging.
+	debugPackage = false
+
+	letters = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
+)
+
+// Tree is a tree.
+type Tree map[string]interface{}
+
+// String implements the Stringer interface method.
+func (t Tree) String() string {
+	y, err := yaml.Marshal(t)
+	if err != nil {
+		return err.Error()
+	}
+	return string(y)
+}
+
+// RandomString returns a random string of length n.
+func RandomString(n int) string {
+	b := make([]rune, n)
+	for i := range b {
+		b[i] = letters[rand.Intn(len(letters))]
+	}
+	return string(b)
+}
+
+// PrettyJSON returns the input JSON properly indented.
+func PrettyJSON(b []byte) []byte {
+	var out bytes.Buffer
+	err := json.Indent(&out, b, "", "  ")
+	if err != nil {
+		return []byte(fmt.Sprint(err))
+	}
+	return out.Bytes()
+}
+
+// IsYAMLEqual compares two YAML string representations by converting them to JSON first.
+func IsYAMLEqual(a, b string) bool {
+	if strings.TrimSpace(a) == "" && strings.TrimSpace(b) == "" {
+		return true
+	}
+	ajb, err := yaml.YAMLToJSON([]byte(a))
+	if err != nil {
+		dbgPrint("bad YAML in isYAMLEqual:\n%s", a)
+		return false
+	}
+	bjb, err := yaml.YAMLToJSON([]byte(b))
+	if err != nil {
+		dbgPrint("bad YAML in isYAMLEqual:\n%s", b)
+		return false
+	}
+
+	return string(ajb) == string(bjb)
+}
+
+// YAMLDiff returns the diff between two YAML string representations.
```

Comment:
In IsYAMLEqual function it do comparison by converting them to JSON first but in YAMLDiff it unmarshall to map[string]interface{} first and marshal then compare. What is the reason for difference behind them?